doc/proposals/fast_ipam_by_node.md

```diff
@@ -0,0 +1,271 @@
+
+Whereabouts currently uses a single lease per cluster named "whereabouts" for locking for all allocation and deallocation
+of IPs across the entire cluster running whereabouts. This causes issues with performance and reliability at scale. 
+Even at 256 nodes, if you have 10 network-attachment-definitions per pod and run a pod on each node there will be so much lease contention that kubelet times
+out before whereabouts can assign all 10 IPs per pod. This would only get worse at higher scale and whereabouts should be able to support
+10+ network-attachment-definition per pod at the kubernetes supported scale of 5,000 nodes.
+
+### Goals
+
+- Support existing whereabouts functionality without breaking changes
+- Introduce a new mode that can be configured on NAD to use IPAM by node slices
+- Support multiple NADs on same network range
+
+### Non-Goals
+
+- Migrate all users to this new mode
+
+<hr>
+
+## Design
+
+The IPAM configuration format would be updated to include enablement of this feature and configurations for the feature.
+
+We will create a new CRD `NodeSlicePool` which will be used to manage the slices of the network ranges that nodes
+are assigned to. A controller where reconcile these NodeSlicePools based on cluster nodes and network-attachment-definitions.
+
+Whereabouts binary will be able to tell this feature is enabled and when creating `IPPools` it will check the `NodeSlicePool` to get the range of the current node.
+It will set this on existing IPPools object and use a lease per IPPool. There will be an `IPPool` and `Lease` per network per node.
+Where a network is defined by network name i.e. you can have multiple `network-attachment-definitions` with a shared network name and this will result in
+a shared `NodeSlicePool`, `IPPool` and `Lease` per node for these `network-attachment-definitions`.
+
+i.e. we have nad1 and nad2 both with network name `test-network`. When a node, `trusted-otter` joins the cluster this will result in
+`NodeSlicePool`, `IPPool` and `Lease` objects named `test-network-trusted-otter`. If these are seperate network you would just not set the network name
+or set the network-name differently per `network-attachment-defintion`.
+
+### Changes in IPAM Config
+
+This will lead to change in IPAM config something as follows:
+
+<table>
+<tr>
+<th>Old IPAM Config</th>
+<th>Changes</th>
+<th>New IPAM Config</th>
+</tr>
+<tr>
+<td>
+  
+```json
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28"
+      }
+}
+```
+  
+</td>
+<td>
+
+```diff
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28"
++       "fast_ipam": true,
++       "node_slice_size": "/22"
```

Comment:
shouldn't the ranges be the other way around ? IIUC, you have a /22 range per node, but the range overall is /28 ...

I would envision something like a /16 range spread across nodes via /24 subnets - i.e. `"range": "192.168.0.0/16"` and `"node_slice_size": "/24"` .

Point being, the mask size on the range must be bigger than the node slice size.

Am I missing something ?

Reply:
bad example for copy pasta another proposal, yes this current configuration would not work because node_slice_size is too big, I'll update with both the comments

Reply:
fixed